src/x/ident/types.go

```diff
@@ -225,6 +225,12 @@ type TagIterator interface {
 	Rewind()
 }
 
+// TagMetadataResolver represents a metadata resolver.
+type TagMetadataResolver interface {
+	// Resolve resolves doc metadata.
+	Resolve(id ID) (doc.Metadata, error)
+}
```

Comment:
The one thing that could reduce allocations even further would be to make this a struct instead of an interface type (i.e. `type TagMetadataResolver struct`) and then just have special constructors for each type and based on which constructor was used set some type of boolean or enum on the struct so that when `Resolve(id) (doc.Metadata, error)` is called it knows what to do.

For an example see the following: `NewMetadataFromIDAndTagIterator` and `NewMetadataFromIDAndTags` and then see the `ResetOrReturnProvidedTagIterator` method.

That way it can be passed on the stack and doesn't need to be heap allocated (which is required with an interface).